Actions for building docker images, alternative model

jashapiro opened the pull request:
This PR is an alternative approach to that taken in #459.

Rather than have one workflow that builds and pushes all of the docker images, with separate workflows for each individual module to test the build only, this keeps both build only and build-push workflows together for each module.
To build all modules, we then have a separate workflow that calls all the others.

The logic in this version is substantially simpler, as each workflow determines when it is triggered, so we save the step of having to determine for ourselves which modules have been modified: if the trigger files are modified, the workflow gets triggered.

The one disadvantage is that while the matrix workflow could incorporate up to 256 individual docker images, the workflow job here can only do 20, so we would potentially need multiple workflows.

There is an alternative here though, which I just thought of: we could move the build-push job out to its own workflow, then have the run-all version use that as well, as a matrix job. Assuming that all workflows have the same steps, that ought to work. I'm going to stack that version onto this one.

jaclyn-taroni commented on .github/workflows/docker_all-modules.yml:
```diff
@@ -0,0 +1,36 @@
+      fail-fast: false
+      matrix:
+        module:
+          - simulate-sce
```
So we'd add to this list as we add modules, correct?

jaclyn-taroni commented on .github/workflows/docker_all-modules.yml:
```diff
@@ -0,0 +1,36 @@
+#
+# This workflow will run only on manual triggers
+
+name: Build all Docker images
+on:
+  workflow_dispatch:
+    inputs:
+      push-ecr:
+        description: "Push to AWS ECR"
+        type: boolean
+        required: false
```
Should this also be scheduled?

jaclyn-taroni replied:
If it should be scheduled, we should probably figure out how we'll know about it if it fails!

jaclyn-taroni commented on .github/workflows/docker_simulate-sce.yml:
```diff
@@ -23,59 +19,40 @@ on:
-    name: Build Docker Image
+  test-build:
+    name: Test Build Docker Image
+    if: github.event_name == 'pull_request' || (contains(github.event_name, 'workflow_') && !inputs.push-ecr)
```
Do I have the right idea: If you're manually calling it or reusing a workflow and push is false, skip straight to `build-push`

jashapiro replied:
If you are manual or reuse and `push-ecr` is false, we run this, if `push-ecr` is true, we skip to `build-push`. (I made the variable different because `push` is overloaded)

jaclyn-taroni commented on .github/workflows/docker_simulate-sce.yml:
```diff
@@ -23,59 +19,40 @@ on:
+    uses: ./.github/workflows/build-push-docker-module.yml
+    if: github.event_name == 'push' || inputs.push-ecr
+    with:
+      module: simulate-sce
```
This field would get populated at template time, yea?

jaclyn-taroni replied:
Looks like yes from what you just pushed!

jashapiro replied:
yup!

jaclyn-taroni commented on .github/workflows/docker_all-modules.yml:
```diff
@@ -0,0 +1,38 @@
+      fail-fast: false
+      matrix:
+        # add all modules with active docker images here
+        module:
+          - hello-python
+          - simulate-sce
```
Here's where we'd keep adding modules, correct? (This was kind of answered by your reorg that wiped out a comment 😄 )

jashapiro replied:
Yes, I added a comment here too because I knew it would be unclear. I'd add another comment at the top to tell people where to go.

jaclyn-taroni commented on .github/workflows/build-push-docker-module.yml:
```diff
@@ -0,0 +1,59 @@
+        id: create-ecr
+        run: |
+          aws ecr-public describe-repositories --repository-names ${{ inputs.module }} \
+          || aws ecr-public create-repository --repository-name ${{ inputs.module }}
+      - name: Set up Docker Buildx
```
```suggestion
          || aws ecr-public create-repository --repository-name ${{ inputs.module }}
          
      - name: Set up Docker Buildx
```

jaclyn-taroni reviewed (commented):
Returning some comments to make sure I'm understanding correctly. 

We should also figure out what kind of interim documentation we need before this goes in (probably stacked on this)

jashapiro commented:
I added filing an issue of failure and adjusted some of the scheduled job timings so that the docker build would run before the analysis module runs, and offset the spelling. 

I had a thought about having the docker build trigger analysis builds when it was done, but that would mean we would run into the max number of workflows with no way to override by adding more actions, so I abandoned that idea.

jaclyn-taroni commented on .github/cron-issue-templates/all-docker-issue-template.md:
```diff
@@ -0,0 +1,7 @@
+At least one of the docker images that is part of the [Build all Docker images](https://github.com/AlexsLemonade/OpenScPCA-analysis/actions/workflows/docker_all-modules.yml) workflow has failed.
```
```suggestion
At least one of the Docker images that is part of the [Build all Docker images](https://github.com/AlexsLemonade/OpenScPCA-analysis/actions/workflows/docker_all-modules.yml) workflow has failed.
```

jaclyn-taroni commented on .github/cron-issue-templates/all-docker-issue-template.md:
```diff
@@ -0,0 +1,7 @@
+https://github.com/AlexsLemonade/OpenScPCA-analysis/actions/workflows/docker_all-modules.yml
+
+Then assign an OpenScPCA admin to this issue and add to this issue a list of the modules with failing docker builds.
+Alternatively, file a new issue for each failing docker image.
```
```suggestion
Alternatively, file a new issue for each failing Docker image.
```

jaclyn-taroni commented on .github/workflows/docker_all-modules.yml:
```diff
@@ -0,0 +1,67 @@
+# This is a workflow to build and optionally push all module Docker images
+#
+# The action works by use a GitHub matrix to build all images in parallel.
+# The modules that will have their docker images built are specified in the matrix variable `module`.
```
```suggestion
# The modules that will have their Docker images built are specified in the matrix variable `module`.
```

jaclyn-taroni commented on .github/workflows/build-push-docker-module.yml:
```diff
@@ -0,0 +1,63 @@
+name: Build and push a docker image for a module
```
```suggestion
name: Build and push a Docker image for a module
```

jaclyn-taroni commented on .github/workflows/docker_simulate-sce.yml:
```diff
@@ -1,8 +1,9 @@
 name: Build docker image for simulate-sce
```
```suggestion
name: Build Docker image for simulate-sce
```

jaclyn-taroni reviewed (approved):
Returning little copy things, but LGTM otherwise 👍🏻
